# Exercise 05 – Cosine of the Angle Between Vectors

In this exercise we compute the cosine of the angle $\theta$ between two vectors $u$ and $v$ using the formula:

$$\cos(\theta) = \frac{u \cdot v}{||u|| \cdot ||v||}$$

This function uses the dot product and the Euclidean norm of the vectors. Note that the behavior is undefined if one or both vectors are zero, or if the vectors have mismatched dimensions.

In [1]:


import sys
# This is the path to the directory where your modules are located in the container
# in this case, it is the directory where the vector.py and matrix.py files are located
# the kernel used here is the one used in the container which is a jupyter kernel
sys.path.append('/home/cliuser/downloads/ft_matrix/ex05')

from vector import Vector

# Example 1: Two identical vectors
u = Vector.from_list([1.0, 0.0])
v = Vector.from_list([1.0, 0.0])
print("Example 1:", "cosine:", Vector.angle_cos(u, v))  # Expected: 1.0

# Example 2: Orthogonal vectors
u = Vector.from_list([1.0, 0.0])
v = Vector.from_list([0.0, 1.0])
print("Example 2:", "cosine:", Vector.angle_cos(u, v))  # Expected: 0.0

# Example 3: Opposing directions
u = Vector.from_list([-1.0, 1.0])
v = Vector.from_list([1.0, -1.0])
print("Example 3:", "cosine:", Vector.angle_cos(u, v))  # Expected: -1.0

# Example 4: Collinear vectors
u = Vector.from_list([2.0, 1.0])
v = Vector.from_list([4.0, 2.0])
print("Example 4:", "cosine:", Vector.angle_cos(u, v))  # Expected: 1.0

# Example 5: General 3D vectors
u = Vector.from_list([1.0, 2.0, 3.0])
v = Vector.from_list([4.0, 5.0, 6.0])
print("Example 5:", "cosine:", Vector.angle_cos(u, v))  # Expected: ~0.974631846

Example 1: cosine: 1.0
Example 2: cosine: 0.0
Example 3: cosine: -0.9999999999999998
Example 4: cosine: 0.9999999999999998
Example 5: cosine: 0.9746318461970762


# Exercise 05 – Cosine of the Angle Between Vectors

In this exercise we compute the cosine of the angle $\theta$ between two vectors $u$ and $v$ using the formula:

$$\cos(\theta) = \frac{u \cdot v}{||u|| \cdot ||v||}$$

This function uses the dot product and the Euclidean norm of the vectors. Note that the behavior is undefined if one or both vectors are zero, or if the vectors have mismatched dimensions.

### Explanation

- **Example 1:** Identical vectors yield $\cos(0°) = 1$
    - When vectors point in the same direction, their angle is 0°
    
- **Example 2:** Perpendicular vectors yield $\cos(90°) = 0$
    - Orthogonal vectors have no directional correlation
    
- **Example 3:** Vectors with opposite directions yield $\cos(180°) = -1$
    - Complete directional opposition
    
- **Example 4:** Collinear vectors (different magnitudes) have $\cos(0°) = 1$
    - Direction matters, not magnitude
    
- **Example 5:** For 3D vectors [1,2,3] and [4,5,6], $\cos(\theta) \approx 0.9746$
    - These vectors are strongly aligned

The cosine similarity provides a measure of directional alignment regardless of vector magnitude.